### ISSUE TITLE
Handle winit's DPI Change event

### ISSUE THREAD

**Issue by jwilm** [jwilm/alacritty#964]
_winit_ added `WindowEvent::HiDPIFactorChanged(f32)`, but none of the backends emit it, yet. Alacritty will need to recompute font size based on the new DPI and flush the glyph cache when the new event is delivered.

This is currently blocked on the event being supported on macOS and X11 _winit_ backends.

cc #510

**Comment by mscharley** [comment]
Why block this on upstream sending events? If the event itself already exists couldn’t you respond to them already and then when they start sending events everything just works and we get a bug fix for something a bit earlier?

**Comment by Strayer** [comment]
According to https://github.com/tomaka/winit/issues/252 the macOS backend should already support the event, but the wayland PR (https://github.com/tomaka/winit/pull/341) seems to be stuck and X11 seems to have issues (https://github.com/tomaka/winit/issues/169).

**Comment by sodiumjoe** [comment]
I just read through the thread and took a look at the code base, and afaict, macOS backend does not emit `HiDPIFactorChanged` (none of them do).

```bash
$ rg HiDPIFactorChanged
CHANGELOG.md
23:- Added event `WindowEvent::HiDPIFactorChanged`.

src/events.rs
103:    HiDPIFactorChanged(f32),
```

**Comment by mathstuf** [comment]
There's a feature matrix that mentions a "new HiDPI" API that has no known support (macOS is an "unknown").

**Comment by thomasjachmann** [comment]
There also is a "HiDPI support" feature that is (only) supported by macOS. Couldn't this feature be used and as soon as the other platforms support it, alacritty's support will increase as well?

**Comment by sodiumjoe** [comment]
The "HiDPI support" row in that matrix does not include `HiDPIFactorChanged` event. It's captured in the "Pending API Reworks" subsection, and it's not implemented for any of the backends.

**Comment by sodiumjoe** [comment]
I have a PR into winit implementing the event for macos: https://github.com/tomaka/winit/pull/443

**Comment by myfreeweb** [comment]
And handling the event could look something like this: https://github.com/myfreeweb/alacritty/commit/3443033dec3b0db909f83f0019c15b6161784da8

(Because I'm working on [a GTK frontend](https://github.com/myfreeweb/galacritty), this is on top of PR #1023 — depends on that Window/Display refactoring. I guess it's not completely impossible to rebase though.)

@sodiumjoe if you haven't reimplemented the same support as I did, you can test my branch + your winit PR :)

**Comment by sodiumjoe** [comment]
@myfreeweb sweet, i'll try that out shortly!

**Comment by sodiumjoe** [comment]
@myfreeweb unfortunately, doesn't seem to have done the trick.

I added a println in your branch

```rust
                    HiDPIFactorChanged(dpr) => {
                        println!("dpi changed: {:?}", dpr);
```

and it logs the correct value when I move the window to another screen, but it doesn't appear to update the font size.

**Comment by sodiumjoe** [comment]
https://github.com/tomaka/winit/pull/443 is now merged, but it still has to be released. I haven't had time to mess around in @myfreeweb 's branch to try to actually get font resizing working though. Hopefully I can find some time this week.

**Comment by sodiumjoe** [comment]
I have a working branch, adapted from @myfreeweb's work, which simply replaces the glyph cache on a dpr change. I'll make a PR as soon as winit and glutin release the upstream fix. I've made a release PR for winit: https://github.com/tomaka/winit/pull/468

**Comment by sodiumjoe** [comment]
Here's the branch in case anyone wants to test/try it out: https://github.com/sodiumjoe/alacritty/tree/964-dpi-change

**Comment by sodiumjoe** [comment]
The upstream change is merged and released in winit. Now just waiting for a bump and release in glutin: https://github.com/tomaka/glutin/pull/1015

**Comment by kwstannard** [comment]
@sodiumjoe 

MacOS 10.12.6
branch: 964-dpi-change

deps:
  glutin: master
  wint: 0.12.0

Loads but moving the window across a DPI change still results in wrong text size.

deps:
  glutin: wint-13
  wint: 0.13.1

Crashes on load.

**Comment by sodiumjoe** [comment]
the upstream change wasn't introduced until `0.13.0`, so the first one is expected.

I'll need more information to diagnose the second one. I would recommend using head of current master of winit, since it includes a bugfix for window titlebars.

What does "crashes" mean? Is there a trace? Can you try with debugging and trace enabled?

**Comment by sodiumjoe** [comment]
@kwstannard `RUST_BACKTRACE=1 cargo run` and `RUST_LOG=alacritty=debug cargo run` might give us more to go on.

**Comment by kwstannard** [comment]
@sodiumjoe Thanks for helping with this.

https://gist.github.com/kwstannard/d0fa4e71ac8a584199115c123a0a5f5f

**Comment by kwstannard** [comment]
These are the alacritty related logs in console:
```
[alacritty pid:65280 mux-aware] starting, non-mux-aware app count 0, switch-state 0 IG FBs 0 EG FBs a:1f power-state 0 3D busy HDA busy
[alacritty pid:65280 non-mux-aware] starting, non-mux-aware app count 1, switch-state 0 IG FBs 0 EG FBs a:1f power-state 0 3D busy HDA busy
[alacritty pid 65280 mux-aware] exiting, non-mux-aware app count 1, runtime: 0:00:00.135
[alacritty pid 65280 non-mux-aware] exiting, non-mux-aware app count 0, runtime: 0:00:00.135
```

**Comment by sodiumjoe** [comment]
@kwstannard ah yes, one of the events changed its name. I just added a commit to my branch that handles that. Can you try again?

**Comment by kwstannard** [comment]
It doesn't immediately close now. I will respond again when I get to a place to test the dpi change.

**Comment by francesca64** [comment]
glutin 0.15.0 is out. @sodiumjoe, I assume you'll be making the PR to update here? If so, [this function](https://github.com/jwilm/alacritty/blob/6debc4f3351446417d0c4e38173cd9ef0faa71d5/src/window.rs#L324) is no longer needed with the new winit version (and is potentially harmful).

**Comment by sodiumjoe** [comment]
@francesca64 Yup, I'll try to get to that tonight. Thanks for the heads-up!

**Comment by sodiumjoe** [comment]
@francesca64 I think I'd actually like to wait for the next version of `winit` so we can include https://github.com/tomaka/winit/pull/488 and https://github.com/tomaka/winit/pull/487, otherwise we would be introducing new known regressions into alacritty.

Also, I don't understand why that function should be removed. Can you elaborate on that?

**Comment by francesca64** [comment]
@sodiumjoe ever since tomaka/winit#451, XIM works out of the box. In fact, while developing that PR (and all PRs since then)  I specifically tested it against Alacritty with that function removed. Additionally, calling `XSetLocaleModifiers` in application code is now *strongly* discouraged, since it can influence Xlib's internal state in surprising ways, potentially interfering with winit's XIM handling. I'm sorry to throw X11 info at you (and clog up this thread), but I see removing it as a necessary part of updating. On that note, keep in mind that until this update lands, Alacritty is still borderline unusable for some users: #830 

Of course, I don't mean to give you a hard time over not wanting to introduce regressions. winit 0.14.0 is still expected to be released in a few days, and tomaka is back, so waiting seems quite reasonable.

**Comment by sodiumjoe** [comment]
@francesca64 Ah ok, that makes complete sense, I'll make sure to add that to my PR when I make it. Meanwhile, I'll keep an eye out for winit `0.14.0`. Thanks for the info!
